compiler/rustc_mir/src/transform/large_enums.rs

```diff
@@ -0,0 +1,272 @@
+                        if matches!(tag_encoding, TagEncoding::Niche { .. }) =>
+                    {
+                        None
+                    }
+                    Variants::Multiple { variants, .. } => {
+                        let min = variants.iter().map(|v| v.size).min().unwrap();
+                        let max = variants.iter().map(|v| v.size).max().unwrap();
+                        if max.bytes() - min.bytes() < D {
+                            return None;
+                        }
+                        let mut discr_sizes = vec![Size::ZERO; adt_def.discriminants(tcx).count()];
+                        for (var_idx, layout) in variants.iter_enumerated() {
+                            let disc_idx =
+                                adt_def.discriminant_for_variant(tcx, var_idx).val as usize;
+                            assert_eq!(discr_sizes[disc_idx], Size::ZERO);
+                            discr_sizes[disc_idx] = layout.size;
+                        }
+                        Some((layout.size, variants.len() as u64, discr_sizes))
+                    }
+                }
+            }
+            _ => None,
+        }
+    }
+    fn optim(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        let mut alloc_cache = FxHashMap::default();
+        let body_did = body.source.def_id();
+        let (bbs, local_decls) = body.basic_blocks_and_local_decls_mut();
+        for bb in bbs {
+            bb.expand_statements(|st| {
+                match &st.kind {
+                    StatementKind::Assign(box (
+                        lhs,
+                        Rvalue::Use(Operand::Copy(rhs) | Operand::Move(rhs)),
+                    )) => {
+                        let ty = lhs.ty(local_decls, tcx).ty;
+
+                        let source_info = st.source_info;
+                        let span = source_info.span;
+
+                        let (total_size, num_variants, sizes) =
+                            if let Some(cand) = Self::candidate(tcx, ty, body_did) {
+                                cand
+                            } else {
+                                return None;
+                            };
+                        let adt_def = ty.ty_adt_def().unwrap();
+
+                        let alloc = if let Some(alloc) = alloc_cache.get(ty) {
+                            alloc
+                        } else {
+                            let mut data =
+                                vec![0; std::mem::size_of::<usize>() * num_variants as usize];
+
+                            let mut curr = 0;
+                            for byte in sizes
+                                .iter()
+                                .flat_map(|sz| sz.bytes().to_ne_bytes())
+                                .take(data.len())
+                            {
+                                data[curr] = byte;
+                                curr += 1;
+                            }
```

Comment:
Check the size of `ptr_sized_integer` to determine how many bytes to write and `data_layout.endian` to select between `to_be_bytes` and `to_le_bytes` (related code in the interpreter would be `write_scalar` & `write_target_uint`).